### PR TITLE
Ziemleszcz/ipsec

### PR DESCRIPTION
**Pull request by ziemleszcz**
## Description

ipsec: pack sadb structs
socket: add socketpair syscall
socket: add sendmsg and recvmsg syscalls
syslog: send messages in the proper way
errno: remove error code duplicates (from phoenix/errno.h)

## Motivation and Context

Changes required for openiked port, mainly unix sockets features and fixes found on the way.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## How Has This Been Tested?

- [ ] Already covered by automatic testing.
- [ ] New test added: (add PR link here).
- [x] Tested by hand on: (imx6ull).

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing linter checks and tests passed.
- [ ] My changes generate no new compilation warnings for any of the targets.

## Special treatment

https://github.com/phoenix-rtos/phoenix-rtos-kernel/pull/228

- [x] This PR needs additional PRs to work (list the PRs, preferably in merge-order).
- [x] I will merge this PR by myself when appropriate.

### CODE REVIEW

File: syslog/syslog.c [hunk 1]
```diff
@@ -111,7 +117,7 @@ void vsyslog(int priority, const char *format, va_list ap)
 	syslog_common.buf[len] = '\0';
 
 	if (syslog_common.open)
-		write(syslog_common.logfd, syslog_common.buf, len + 1);
+		sendto(syslog_common.logfd, syslog_common.buf, len + 1, 0, (struct sockaddr *)&syslog_common.addr, syslog_common.addrlen);
```
Comment (nalajcie) [inline syslog/syslog.c:120]: why not `write`?

File: syslog/syslog.c [hunk 2]
```diff
@@ -58,13 +60,17 @@ void openlog(const char *ident, int logopt, int facility)
...
 		syslog_common.ident = argv_progname;
 	}
 
 	if (logopt & LOG_NDELAY) {
 		if (!syslog_common.open) {
-			syslog_common.logfd = open(PATH_LOG, O_WRONLY | O_NONBLOCK);
+			syslog_common.addr.sun_family = AF_UNIX;
+			strcpy(syslog_common.addr.sun_path, PATH_LOG);
+			syslog_common.addrlen = SUN_LEN(&syslog_common.addr);
+			syslog_common.logfd = socket(AF_UNIX, SOCK_DGRAM | SOCK_NONBLOCK, 0);
```
Comment (nalajcie) [inline syslog/syslog.c:73]: maybe `connect` once and then use `write`? I don't know the implementation but it should be faster?

**Review by nalajcie — COMMENTED** [review]
Left some comments, we will still need to wait for kernel PR to pass before merging this.

The changes seem sane.

File: sys/socket.c [hunk 3]
```diff
@@ -65,99 +72,150 @@ ssize_t send(int socket, const void *message, size_t length, int flags)
...
+				.msg_namelen = msg->msg_namelen,
+				.msg_iov = &_iov,
+				.msg_iovlen = 1,
+				.msg_control = msg->msg_control,
+				.msg_controllen = msg->msg_controllen,
+				.msg_flags = msg->msg_flags
+			};
+
+			if (len <= 64) {
+				char buf[len]; /* use VLA for small buffers */
```
[the remark this answers is not in the document]
Comment (ziemleszcz) [reply sys/socket.c:146]: OK, I thought we support C99. It's already 2021, how much longer do we have to wait?

File: sys/socket.c [hunk 4]
```diff
@@ -373,80 +433,6 @@ void freeaddrinfo(struct addrinfo *res)
 }
 
 
-int socketpair(int domain, int type, int protocol, int socket_vector[2])
```
[the remark this answers is not in the document]
Comment (ziemleszcz) [reply sys/socket.c:376]: Explained in kernel pull request.

File: syslog/syslog.c (hunk 2 above)
Comment (ziemleszcz) [reply syslog/syslog.c:73]: Connect may fail if syslogd socket is not available yet (syslogd starting). What if syslogd restarts or crashes? It safer to always send to destination address.

File: syslog/syslog.c (hunk 1 above)
Comment (ziemleszcz) [reply syslog/syslog.c:120]: Check comment below.

File: sys/socket.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (nalajcie) [reply sys/socket.c:146]: We do support C99, but our coding guidelines say that we should use `C89` so  until we change our coding style we should stick with it (even as much as I would like to use `C99` too).

Actually for VLA the situation is shaky as it was introduced in C99 as standard feature and than moved to optional feature in C11 (so compiler might not support it)

File: sys/socket.c [hunk 5]
```diff
@@ -65,99 +72,150 @@ ssize_t send(int socket, const void *message, size_t length, int flags)
...
+
+				_iov.iov_base = buf;
+				copy_from_iov(buf, msg->msg_iov, msg->msg_iovlen);
+				len = sys_sendmsg(socket, &_msg, flags);
+			}
+			else {
+				void *buf;
+
+				buf = malloc(len);
+				if (!buf)
```
Comment (agkaminski) [inline sys/socket.c:156]:
```suggestion
				if (buf == NULL)
```

File: sys/socket.c [hunk 6]
```diff
@@ -65,99 +72,150 @@ ssize_t send(int socket, const void *message, size_t length, int flags)
...
+
+				_iov.iov_base = buf;
+				len = sys_recvmsg(socket, &_msg, flags);
+				copy_to_iov(buf, msg->msg_iov, msg->msg_iovlen, len);
+			}
+			else {
+				void *buf;
+
+				buf = malloc(len);
+				if (!buf)
```
Comment (agkaminski) [inline sys/socket.c:204]:
```suggestion
				if (buf == NULL)
```

**Review by agkaminski — APPROVED** [review]
Quite nice idea for elimination of alignof

**Comment by nalajcie** [comment]
merging at the same time as changes in phoenix-rtos-kernel, CI should pass on submodules update

**Pull request merged**